### ISSUE TITLE
Some errors when using multi-GPU

### ISSUE THREAD

**Issue by beichao1314** [OpenNMT/OpenNMT#58]
I try to train by using multi-GPU. When I test it, there are some problems:
1. When options **-nparallel 2** and **-gpuid 14** are set, the first GPU and the second GPU are used.(I have 16 GPUs.) So I can't set the specific GPU when using multi-GPU. The first and the second GPU are running other programs. And here is error:

Loading data from './data/test-train.t7'...	
 * vocabulary size: source = 10002; target = 10004	
 * additional features: source = 0; target = 0	
 * maximum sequence length: source = 50; target = 51	
 * number of training sentences: 9453	
 * maximum batch size: 16	
Building model...	
 * using input feeding
Initializing parameters...
 * number of parameters: 28777004
Preparing memory optimization...
 * sharing 70% of output/gradInput tensors memory between clones
Start training...	
	
/home/beichao/torch/install/bin/lua: ./onmt/train/Optim.lua:83: Assertion `THCTensor_(checkGPU)(state, 1, self)' failed.  at /home/beichao/torch/extra/cutorch/lib/THC/generated/../generic/THCTensorMathReduce.cu:180
stack traceback:
[C]: in function 'norm'
./onmt/train/Optim.lua:83: in function 'prepareGrad'
train.lua:294: in function 'trainEpoch'
train.lua:411: in function 'trainModel'
train.lua:548: in function 'main'
train.lua:551: in main chunk
[C]: in function 'dofile'
...chao/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
[C]: in ?

2. When options **-nparallel 2**, **-gpuid 14** and **-async_parallel** are set, the first GPU and the second GPU are still used. It can run first epoch but when saving the model, there is an error:

Loading data from './data/test-train.t7'...	
 * vocabulary size: source = 10002; target = 10004	
 * additional features: source = 0; target = 0	
 * maximum sequence length: source = 50; target = 51	
 * number of training sentences: 9453	
 * maximum batch size: 16	
Building model...	
 * using input feeding
Initializing parameters...
 * number of parameters: 28777004
Preparing memory optimization...
 * sharing 70% of output/gradInput tensors memory between clones
Start training...	
	
Epoch 1 ; ... batch 50/610
Epoch 1 ; ... batch 100/610
Epoch 1 ; ... batch 150/610
Epoch 1 ; ... batch 200/610
Epoch 1 ; ... batch 250/610
Epoch 1 ; ... batch 300/610
Epoch 1 ; ... batch 350/610
Epoch 1 ; ... batch 400/610
Epoch 1 ; ... batch 450/610
Epoch 1 ; ... batch 500/610
Epoch 1 ; ... batch 550/610
Epoch 1 ; ... batch 600/610
Epoch 1 ; Iteration 612/610 ; Learning rate 1.0000 ; Source tokens/s 620 ; Perplexity 10669.35	
Saving checkpoint to 'en2zh_test_model_checkpoint.t7'...	
/home/beichao/torch/install/bin/lua: ./onmt/utils/Tensor.lua:81: Assertion `THCTensor_(checkGPU)(state, 1, self_)' failed.  at /home/beichao/torch/extra/cutorch/lib/THC/generic/THCTensorMath.cu:21
stack traceback:
[C]: in function 'zero'
./onmt/utils/Tensor.lua:81: in function 'reuseTensor'
./onmt/utils/Tensor.lua:120: in function 'initTensorTable'
./onmt/modules/BiEncoder.lua:118: in function 'forward'
train.lua:166: in function 'eval'
train.lua:413: in function 'trainModel'
train.lua:548: in function 'main'
train.lua:551: in main chunk
[C]: in function 'dofile'
...chao/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
[C]: in ?

In a word, I meet the same problems when I train by synchronous parallelism and asynchronous parallelism, **can't specify the GPUs** and **Assertion `THCTensor_(checkGPU)(state, 1, self_)' failed**. Anyone can help me? Thanks anyway.

**Comment by guillaumekln** [comment]
Yes that is an issue if GPUs can't be set.

@jsenellart-systran Maybe we should drop the `-nparallel` option and support a list of comma-separated GPU identifiers with `-gpuid`. What do you think?

**Comment by jsenellart-systran** [comment]
actually, the problem can be fixed by using CUDA_VISIBLE_DEVICES=xx,yy and then set -gpuid 1 (the 1 here will be xx, and the second GPU used with be yy)

In any case this method is preferable otherwise you will be using some memory on all GPUs.

@guillaumekln - I do like however, your suggestion too - so we can drop `-nparallel` that is replaced by the list on the `-gpuid` option. but here also we need to combine with `CUDA_VISIBLE_DEVICES` (except if we can programmatically do that before loading cutorch?

**Comment by guillaumekln** [comment]
Right. Then dropping `-nparallel` is not so nice because GPU identifiers need to be listed in `-gpuid` and `CUDA_VISIBLE_DEVICES`.

I will document the use of `CUDA_VISIBLE_DEVICES` in the guide.

**Comment by beichao1314** [comment]
Thanks！
I just set `-gpuid 1`, and there is no error. But if gpuid > 1, the first and the second gpu are used, then the error is appeared.

**Comment by jsenellart-systran** [comment]
I am not sure to understand, I recommend using `-gpuid 1` and control the actual GPU with `CUDA_VISIBLE_DEVICES` - does this work?

**Comment by beichao1314** [comment]
When I set `-gpuid 1` and control the actual GPU with CUDA_VISIBLE_DEVICES, it works well. But there is an error when I set `-gpuid 2` and `CUDA_VISIBLE_DEVICES=1, 14`. Do I have to set `-gpuid 1` when using multi-GPU?

**Comment by jsenellart-systran** [comment]
Yes - it was designed like that and there is no constraint since you can even order `CUDA_VISIBLE_DEVICES` differently if you need.
More generally, I think we need to think more about the option `-gpuid 1` and how to use it. I will think more about it. Thanks for your testing.

**Comment by beichao1314** [comment]
Thanks for your reply !
